packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,484 @@
+ */
+type HTMLElementWithStyledMap<T extends HTMLElement = HTMLElement> = T & {
+  computedStyleMap(): StylePropertyMapReadOnly;
+};
+
+/**
+ * CSS with number parsing.
+ * @see https://drafts.css-houdini.org/css-typed-om-1/#css
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/CSS/number
+ */
+type CSSWithNumber = typeof CSS & {
+  number(value: number): {
+    value: number;
+    readonly unit: string;
+  };
+};
+
+/**
+ * @internal
+ *
+ * Motion props.
+ */
+export type MotionProps<T = HTMLElement> = {
+  /**
+   * Whether the element should be present in the DOM.
+   *
+   * @default false
+   */
+  presence: boolean;
+
+  /**
+   * Ref to the element.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} />
+   */
+  ref?: RefObjectFunction<T> | React.RefCallback<T> | React.Ref<T>;
+
+  /**
+   * Whether the element is currently active in the DOM.
+   * Useful to apply CSS transitions only when the element is active.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} className={motion.active ? 'element element-active' : 'element'} />
+   */
+  active?: boolean;
+
+  /**
+   * Current state of the element.
+   *
+   * - `unmounted` - The element is not yet rendered or can be safely removed from the DOM.
+   * - `entering` - The element is performing enter animation.
+   * - `entered` - The element has finished enter animation.
+   * - `idle` - The element is currently not animating, but rendered on screen.
+   * - `exiting` - The element is performing exit animation.
+   * - `exited` - The element has finished exit animation.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} className={`element element-${motion.state}`} />
+   */
+  state?: 'unmounted' | 'entering' | 'entered' | 'idle' | 'exiting' | 'exited';
+};
+
+export type MotionOptions = {
+  /**
+   * Whether to animate the element on first mount.
+   *
+   * @default false
+   */
+  animateOnFirstMount?: boolean;
+};
+
+/**
+ * @internal
+ *
+ * Returns CSS styles of the given node.
+ * @param node - DOM node.
+ * @returns - CSS styles.
+ */
+const getElementComputedStyle = (node: HTMLElement): CSSStyleDeclaration => {
+  const win = node.ownerDocument?.defaultView ? node.ownerDocument.defaultView : window;
+
+  if (!win || !canUseDOM()) {
+    return {
+      getPropertyValue: (_: string) => '',
+    } as CSSStyleDeclaration;
+  }
+
+  return win.getComputedStyle(node, null);
+};
+
+/**
+ * Converts a CSS duration string to milliseconds.
+ *
+ * @param duration - CSS duration string
+ * @returns Duration in milliseconds
+ */
+function toMs(duration: string): number {
+  const trimmed = duration.trim();
+
+  if (trimmed.includes('auto')) {
+    return 0;
+  }
+
+  if (trimmed.includes('ms')) {
```

Comment:
I think this better aligns with the expected input.

```suggestion
  if (trimmed.endsWith('ms')) {
```